Test 9806

indio99 opened the pull request:
Hi.

This changes work fine for me.

prenticedavid commented:
It is difficult to compare your current version of CPP with mine.   It looks as if you have got an identical version to me.    But unfortunately GitHub has recorded all sorts of trivial edits.

I suggest that you try different values for OFFSET_9806.   I would expect 10 to work.    It is currently set to 96 which will probably be crap.

I am horrified by your shield.h
Please copy-paste my version to replace yours.

I have shield.h using digitalWrite() for the PIN_LOW macro
Get the display working correctly with  OFFSET_9806 first.

Then try my alternative macros for PIN_LOW

I will assume that you will start with the graphictest_kbv example with my exact library files.
I will be able to see any subsequent edits that you make.

David.

prenticedavid commented:
@indio99,
I think that I have identified the GitHub anomaly.  My CPP file has trailing spaces on several lines.  e.g. the define SUPPORT_8357D_GAMMA line.
When I create a Branch the trailing spaces go with it.

However,  your Fork has removed the trailing spaces.   So when I compare your files with mine,   I get trivial differences due to the trailing spaces.

One day I will re-format all these files to remove tabs and replace with spaces.   And remove those trailing spaces!
That is the price you pay for editing with Notepad++, Keil, Rowley, AS7, Arduino and even Notepad sometimes. 

David.

indio99 commented:
Oh, I’m sorry, I copy and paste from Atom, I edit and compile with platformio.

In cpp I only apply two changes , I added in line 176 

 #ifdef SUPPORT_9806
      //return 0x9806;
      ret = readReg32(0xD3);           //Obtain ID
      //Serial.print("RET ");printhex(ret);Serial.println();
      if (ret == 0x9806){             //Check if 0x9806
          //Serial.println("Return ILI9806");
          //readReg32(0xD3);           //forces a read again, I observed that if not, don't work propperly
          return 0x9806;
        }
    #endif

Add in 2804
(0x3A), 1, /*Tearing Effect ON*/0x55,
This comment are wrong, this is for pixel format

And in shield.h is easy to see, very little get trivial differences by spaces.

Your PinLow and High
(*((volatile uint32_t*)(&p)+2)) = (1<<(b&31))
 work correctly for all pins, but when apply to cd_pin, tft dont start, for this reason I modify this changes

prenticedavid commented:
0x3A is pixfmt.   Anyway,  I set pixfmt as part of the 
``
		init_table(&wake_on, sizeof(wake_on));
``

sequence.   So the comment / command is not important.

We can sort out the CD_COMMAND timing later.   I am a little surprised that the ESP32 does not force me to use a WRITE_DELAY.

Does  OFFSET_9806 really work with a value of 96 ?
Your GitHub CPP seems to still have 96.

David.

indio99 commented:
Hi.

I probe with different values of offset, positive and negative values, but in REV displays, only desplace the screen, the fix black bar allways see it with out changes

prenticedavid commented:
As long as you get different results you can describe them.
e.g. can you get the DEADBEEF to become visible?
Where is it displayed for -32, +10, +32, +64, +96, ... ?

Or use the diagnose_TFT screen to view the white rectangles.   Whichever is easier.

David.

indio99 commented:
Hi.

I used now the diagnose_TFT, and I surprised because now I see that in REV mode don't work fillScreen() and into drawRect() don't work fastVLine() too.

For can I see two bottom lines I need to apply -70 to offset.

![img_5956](https://user-images.githubusercontent.com/11825217/43169621-0f7c99fa-8fa2-11e8-8e12-18d4c2558f30.JPG)

prenticedavid commented:
I have put a diagnosis sketch in the examples.
It should run on your ESP32.   Observe the behaviour when bits in MADCTL are changed.

I am going to Scotland for 4 days.   We can probably sort this out next wee

David.

indio99 commented:
Hi.
Firstly, have you the best travel ;D

I tested your new diagnose sketch, and I attach a new video with the result.

I don't changed any at this time, on your sketch, and on your cpp or shield.h, this video is with out any changes.

[http://www.codetronic.es/owncloud/index.php/s/jQRRjXDgN7tSbo7](url)

prenticedavid commented:
Thanks for the video.   Look at 
0.05s 0:flip vert
0.36s 4:vert refresh
1.00s 7.Row Order

Row Order is behaving as expected i.e. needs OFFSET=10
Vert Refresh normally shows nothing (it affects the vert scroll direction)
Vert Flip is normally similar to Row Order.

I will put an update on GitHub later today.   I have never seen this behaviour on Ilitek controllers before.

Good to come back to some rain in Kent.

David.

indio99 commented:
Hi, excuse me for waiting.

I was try with the last change, I don't modified any things, I download and applied directly.

And this is the result. At end I attach a video with complete sequence.

![img_6092](https://user-images.githubusercontent.com/11825217/43709208-879b44a8-996c-11e8-8bfa-7208a88325ec.JPG)
![img_6093](https://user-images.githubusercontent.com/11825217/43709211-87ce0d20-996c-11e8-9eac-8e25169e3c44.JPG)

I tried with different values of offset (10, 32, 70, -10, -32, -70), but all  I see it same, any differences between values offer same result.

[http://www.codetronic.es/owncloud/index.php/s/7YocvzNXex5eUSC](url)

prenticedavid commented:
That is what I would expect for this sketch.    It draws the "White" picture in Portrait.    As you can remember,  Portrait works perfectly.    The "Blue/Red" pictures show the effect of changing a single bit in the MADCTL register.    It never shows the effect of OFFSET_9806 because that is only used in xxx_Rev rotations.

Did you try the graphictest_kbv sketch?

That displays Portrait, Landscape, Portrait_Rev, Landscape_Rev
And so does diagnose_TFT_controller and readPixel sketch

David.

indio99 commented:
Hi.

Yes you are right. Now work correctly. Good value for offset are 10.

Only I need made minor modifications for work well.

First, in readID(), if I don't check 0xD3 register and return immediately occur this:
[video1](http://www.codetronic.es/owncloud/index.php/s/5itrCSQorZIOSjq)

If I check 0xD3 first on readID() this it see now:
[video2](http://www.codetronic.es/owncloud/index.php/s/KD0ItwtH2b0rbDt)

And in graphictest_kbv.ino, for that it work well, I need to modify timing of writeCmd and writeData, if not modify this, graphic test form adafruit_gfx crash it.

in this video you can see the result with all modifications
[video3](http://www.codetronic.es/owncloud/index.php/s/TZdg6R9HWWy0kQs)  

I go to update all modifications on my fork of this branch for you can see and evaluate.

Thanks

prenticedavid commented:
The diagnose_TFT_support sketch reads the ID in setup().   This ID is reported to Serial Terminal.

readID() is called in every loop().   This is the value that you see on the TFT screen.

Please inspect the Serial Terminal when debugging.    It is using the first ID from setup() which accounts for the wrong colours and geometry.

I am not happy with calling readReg32(0xD3) first in the readID() method.
Please experiment with readReg32(0xD3) where I call it.     Possibly call two or three times to check for 0xD3D3.

Regarding WriteCmd() and WriteData().   I would prefer to stick with digitalWrite() for the PIN_LOW() if it works reliably on your ESP32.    The other macro is too fast.  

The alternative is to introduce a delay:
`#define WRITE_DELAY { WR_ACTIVE; }
` 

So the main job is to experiment with readID()

David.

indio99 commented:
Hi.

I tried to find bad value on readID(), but on wrong colors and geometry, the ID is the same, 0x9806.

I tried to other configs in readID(), but I concluded that first readreg32() called before to 0xD3 are the responsible of corruption.
I put this before firs readReg32(), and voila, work well
`ret = readReg(0xD3, 1);      //for ILI9806`
`if (ret == 0x9806) return ret;`

For the macros of PIN_LOW I go try in next.

indio99 commented:
Hi. I updated my fork branch, you can see modifications.

Basically, in .cpp I only changed position of read RET variable, because if I read any readReg32 before, the colors and geometry are corrupted, other thing that I don't say before, is that when this occur, I need disconnect power line completely from TFT, if not, colors or geometry are always corrupted.

In .h, I found an alternate for WriteCmd and WriteData, but for CD_COMMAND and CD_DATA macros, I don't found an alternate yet.

If I use digitalWrite macros, all screens are very slow, this is not a problem, we are testing, but the real problem is that test graphics based on adafruit_FX crashes extrepitously.

Thanks

prenticedavid commented:
I find it a bit difficult to compare my test_9806 files with your ones.   Due to the trailing space thing. 

I have trimmed my files.   Please can you start again with my current test_9806 branch.
I have set massive delays and defaulted to digitalWrite()

With my slowest shield (ILI9320):
I get reliable results with those settings (5.41s) i.e. WR_ACTIVE4, RD_ACTIVE8 with digitalWrite
I get reliable results (1.68s) i.e. WR_ACTIVE0, RD_ACTIVE2 with memory-cast

I do not understand why you get glitches with digitalWrite.   Or why you need the complex WriteCmd() macro.

My usual strategy for an unknown controller speed is:
1.   readID() and display on Serial.
2.   force the correct ID in tft.begin(ID)
3.   adjust WR_ACTIVE macros until there are no glitches
4.   observe the ID value reported in Adafruit Test Results
5.   adjust RD_ACTIVE macros until readID() works correctly.   Software Scroll is correct.
6.   use the correct ID from readID() in begin(ID)  

I suggest that you copy my files as-is.    Then we are both starting from the same point.    

If you really need your conditional macros,    I can only accept your word.    You know how to add them.
It seems very odd that your hand-wired ESP32 is that much more glitchy than my neat pcb with Arduino headers.

Ah-ha.   I have just tried on my hand-wired ESP32-module Protoshield.    It is a little glitchy with digitalWrite.    It is VERY unreliable with memory-cast.

We are trying to send fast parallel signals from a 240MHz CPU over ratsnest wiring.
I am not surprised that there are difficulties.

David.

indio99 commented:
I'm making a personal shield for this microcontroller and this TFT, my wires are 150mm long, and in the shield, max long of wire will are of 30mm.

<img width="1186" alt="captura de pantalla 2018-08-07 a las 13 50 18" src="https://user-images.githubusercontent.com/11825217/43774771-9909035e-9a4a-11e8-93fd-b6a76c7d4bac.png">


I need route more nets yet, but when I have this shield, I'll can try with it, perhaps glitchy not occur in digitalWrite.

I'll try to correct the branch of fork for you can see correctly.

prenticedavid commented:
My Protoshield is fairly neat and tidy.   The ESP32 module is mounted on the Protoshield with wires soldered from the Arduino headers to the ESP32 module pins.    Obviously some data lines are shorter than others.    The pcb traces on the TFT shield must be 70mm or so.   The wires from Arduino header to ESP32 must be another 70mm.

I suspect that the critical wires are the /WR and /RD lines.    I would expect /RS and /CS to be reasonably tolerant of ringing.    And the Data Bus lines must be stable before /WR is latched.

Regarding the "test_9806" branch.   We need to get a reliable readID()
You have discovered what you require for reliable PIN_LOW and PIN_HIGH.

David.

indio99 commented:
Hi.

I find now that the conflict doesn't with CD_COMMAND, is with CD_DATA.

I added NOP() for void cicles of clock before active output of CD_DATA and it working well, I adjusted the number of NOP to work without glitches.

readID() work well like I uploaded now. Reading readReg(0xD3, 1) before first readReg32.
This is very stable for me.

In my actual shield pcb has about 100mm traces net from pinhead connector to fppc(flexwire) connector, plus 150mm manual wiring, 250mm in total from esp32 to flexwire of TFT. In the pcb that I designing, the total length from esp32 to flexwire of TFT will are about 30mm.

I updated cpp and h files.

prenticedavid commented:
I have never tried NOP with ESP32.    The ARM compiler will optimise GPIO.    That is why I use "volatile" macros like WR_ACTIVE

I suspect that the only critical line is /WR.   If any ringing occurs,  the command or data will get out of step.

You might solve instability problems with a small RC low pass filter on the /WR line e.g. with 10ns time constant.    100R x 100pF = 10ns.    Pure guesswork.   You might get away with just a small capacitor close to the ribbon.

However much ringing occurs on the /RS line (CD_DATA or CD_COMMAND),   I would be 100% confident that it will have decayed by the time /WR is latched.

If I feel enthusiastic,   I might experiment with RC filter on my Protoshield.

David.

indio99 commented:
Hi.

I tried with different RC low pass filter like you recommend me, 100R x 100pf too, I don't see any differences.
I tried RC Low pass filter in WR and RS line, I don't see any differences too.

I continued debuging mcufriend_shield.h and I obtain a setup that is stable for me and about one second faster in the complete sequence of graphictest, from 11.98sec to 11.10sec. I'm not want performance or ultra speed, I want stability and this setting are stable for me.

If I use only digitalWrite Macros, in adafruit_GFX it glitches and crash.
If I use only 'p' Macros, it don't work any.
Only if I use hybrid of 'p' Macros in convination with  CD_DATA modificated, work well.

I updated h file to last modifications in my fork your code, I don't know how to do for integrate CD_DATA macro without alter your code structure. I only known that using #define and #ifndef work, as you can see in my fork of your code, but I don't know other method. I understand that you don't like this.

Respect to readID(), if you don't have inconvenient of this sequence, readReg(0xD3, 1) before the first readReg32(), we can close this. I think that it don't perturb to other TFTs.

prenticedavid commented:
I have not forgotten you.    I will experiment with RC filters.    Possibly tomorrow.

prenticedavid commented:
Sorry,  I still have not experimented with RC filters yet.

Regarding optimisation.   Yes,  I could optimise when high == low.   i.e. just strobe like the 16-bit interface does.
In practice,  you only get high == low for BLACK and WHITE.    It speeds up the Adafruit Tests.   I doubt if you would notice in a regular application.
The ILI9488 can do 8-bit colour.   The ILI9806 can NOT do 8-bit colour. 

David.
